node/src/components/block_proposer/deploy_sets.rs

```diff
@@ -92,6 +105,14 @@ impl BlockProposerDeploySets {
             [pending_deploys, pending_transfers].concat(),
         )
     }
+
+    /// Adds a finalized deploy hash.
+    pub(super) fn add_finalized(&mut self, deploy_hash: DeployHash, new_expiry: Timestamp) {
+        self.finalized_deploys
+            .entry(deploy_hash)
+            .and_modify(|expiry| *expiry = new_expiry.min(*expiry))
```

Comment:
Why would a deploy be finalized more than once?

Reply:
In the situation where we would have used `unhandled_finalized`, we would now add the deploy twice:

If we learn about a finalized block containing a deploy for which we are missing the header, we'll add it to `finalized_deploys` with the latest possible expiry date based on the block timestamp and max TTL.

If we then learn about the deploy's header later, we can update its expiry date and use the correct one, based on the deploy's timestamp and actual TTL.